### PR TITLE
Updater class

### PR DESCRIPTION
**Pull request by JanKoelzer**
When creating a cell, an `Updater` is passed to the constructor (instead of a lattice). This `updater` may
(a) join alls values that get put to the cell ( `AggregationUpater`) or
(b) check, if the new value is a refinement and then simple store it – or otherwise throw ( `MonotonicUpdater`).

We have discussed use cases for both approaches. Most of our test currently use (b).

### CODE REVIEW

**Comment by JanKoelzer** [comment]
(This contains #90 )

File: core/src/main/scala/lattice/Lattice.scala [hunk 1]
```diff
@@ -2,38 +2,69 @@ package lattice
...
+   * r == 0 iff x == y
+   * r > 0 iff x > y
+   */
+  override def tryCompare(x: V, y: V): Option[Int] =
+    if (lt(x, y)) Some(-1)
+    else if (gt(x, y)) Some(1)
+    else if (equiv(x, y)) Some(0)
+    else None
+
+  def empty: V
```
Comment (JanKoelzer) [inline core/src/main/scala/lattice/Lattice.scala]: Rename to `bottom`.

File: core/src/main/scala/lattice/updater.scala [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package lattice
+
+trait Updater[V] {
+  def empty: V
```
Comment (JanKoelzer) [inline core/src/main/scala/lattice/updater.scala]: Use self-type PartialOrderingWithBottom => { def empty: V = bottom }

File: core/src/main/scala/lattice/updater.scala [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+trait AggregationUpdater[V] extends Lattice[V] with Updater[V] {
+  override def update(current: V, next: V): V = join(current, next)
+}
+
+trait FixedUpdater[V] extends PartialOrderingWithBottom[V] with Updater[V] {
+  override def update(current: V, next: V): V =
+    if (lteq(current, next)) next
+    else throw new LatticeViolationException[V](current, next)
```
Comment (JanKoelzer) [inline core/src/main/scala/lattice/updater.scala]: Remove `[V]`.

**Comment by JanKoelzer** [comment]
All requested changes have been made. This should be ready now.

File: core/src/main/scala/lattice/Lattice.scala [hunk 4]
```diff
@@ -2,38 +2,68 @@ package lattice
...
+   * Return the join of v1 and v2 wrt. the lattice.
    */
-  def join(current: V, next: V): V
-  def empty: V
+  def join(v1: V, v2: V): V
+
+  def lteq(v1: V, v2: V): Boolean = {
+    try
+      join(v1, v2) == v2
+    catch { case _: LatticeViolationException[V] => false }
```
Comment (phaller) [inline core/src/main/scala/lattice/Lattice.scala]: I don't think we should expect an exception to be thrown by `join(v1, v2)`, because in a `Lattice[T]` the result of `join(v1, v2)` should always be well-defined. In fact, I think it should be part of the contract of trait `Lattice[T]` that its `join` method should never throw an exception. (Otherwise, we are obviously not implementing a lattice!)

Therefore, I think the `try-catch` should be removed.

File: core/src/main/scala/lattice/Lattice.scala [hunk 5]
```diff
@@ -2,38 +2,68 @@ package lattice
...
+  }
+
+  override def gteq(v1: V, v2: V): Boolean = {
+    try
+      join(v1, v2) == v1
+    catch { case _: LatticeViolationException[V] => false }
```
Comment (phaller) [inline core/src/main/scala/lattice/Lattice.scala]: Ditto.

File: core/src/main/scala/cell/HandlerPool.scala [hunk 6]
```diff
@@ -7,7 +7,9 @@ import scala.annotation.tailrec
 import scala.util.control.NonFatal
 import scala.concurrent.{ Await, Future, Promise }
 import scala.concurrent.duration._
-import lattice.{ DefaultKey, Key, Lattice }
+
+
```
Comment (phaller) [inline core/src/main/scala/cell/HandlerPool.scala]: Why two newlines?

File: core/src/main/scala/cell/HandlerPool.scala [hunk 7]
```diff
@@ -37,23 +39,23 @@ class HandlerPool(parallelism: Int = 8, unhandledExceptionHandler: Throwable =>
    *
    * @param key The key to resolve this cell if in a cycle or no result computed.
    * @param init A callback to return the initial value for this cell and to set up dependencies.
-   * @param lattice The lattice of which the values of this cell are taken from.
+   * @param updater The lattice of which the values of this cell are taken from.
```
Comment (phaller) [inline core/src/main/scala/cell/HandlerPool.scala]: "The lattice of which the values of this cell are taken from." -> "The updater used to update the value of this cell."

File: core/src/main/scala/cell/HandlerPool.scala [hunk 8]
```diff
@@ -37,23 +39,23 @@ class HandlerPool(parallelism: Int = 8, unhandledExceptionHandler: Throwable =>
...
+    CellCompleter(key, init)(updater, this).cell
   }
 
   /**
    * Returns a new cell in this HandlerPool.
    *
    * Creates a new, completed cell with value `v`.
    *
-   * @param lattice The lattice from which the values of this cell are taken
+   * @param updater The lattice from which the values of this cell are taken
```
Comment (phaller) [inline core/src/main/scala/cell/HandlerPool.scala]: "The lattice from which the values of this cell are taken" -> "The updater used to update the value of this cell."

File: core/src/main/scala/cell/Cell.scala [hunk 9]
```diff
@@ -7,6 +7,9 @@ import scala.annotation.tailrec
 import scala.util.{ Failure, Success, Try }
 import lattice.{ DefaultKey, Key, Lattice, LatticeViolationException }
 
+import lattice.{ Updater, LatticeViolationException, Key, DefaultKey }
```
Comment (phaller) [inline core/src/main/scala/cell/Cell.scala]: Integrate imports into line 8.

File: bench/src/test/scala/benchmark/rabenchmarks.scala [hunk 10]
```diff
@@ -24,7 +23,7 @@ object ReactiveAsyncBenchmarks extends PerformanceTest.Microbenchmark {
   val size = Gen.single(s"$nrOfCells cells")(nrOfCells)
 
   /* lattice instance for cells */
-  implicit val naturalNumberLattice: Lattice[Int] = new NaturalNumberLattice
+  implicit val naturalNumberUpdater: Updater[Int] = new NaturalNumberUpdater
```
Comment (phaller) [inline bench/src/test/scala/benchmark/rabenchmarks.scala]: This should stay as before, and instead the implicit conversion from lattice to updater should take care of it.

File: bench/src/test/scala/benchmark/rabenchmarks.scala [hunk 11]
```diff
@@ -1,9 +1,8 @@
 package cell
 
-import lattice.{ Lattice, NaturalNumberLattice, NaturalNumberKey }
+import lattice._
```
Comment (phaller) [inline bench/src/test/scala/benchmark/rabenchmarks.scala]: Please revert this change. Wildcard imports are a suboptimal default, and there is no good reason to prefer a wildcard import in this case.

File: bench/src/test/scala/benchmark/rabenchmarks.scala [hunk 12]
```diff
@@ -1,9 +1,8 @@
...
 
 import scala.annotation.tailrec
-
```
Comment (phaller) [inline bench/src/test/scala/benchmark/rabenchmarks.scala]: If you revert this change, this file would not have to be touched at all. 😃

File: core/src/main/scala/lattice/Lattice.scala [hunk 13]
```diff
@@ -2,38 +2,68 @@ package lattice
...
   def trivial[T >: Null]: Lattice[T] = {
     new Lattice[T] {
-      override def empty: T = null
-      override def join(current: T, next: T): T = {
-        if (current == null) next
-        else throw LatticeViolationException(current, next)
+      override def bottom: T = null
+      override def join(v1: T, v2: T): T = {
+        if (v1 == null) v2
+        else throw LatticeViolationException(v1, v2)
```
Comment (phaller) [inline core/src/main/scala/lattice/Lattice.scala]: The implementation of `join` for a lattice should never throw this exception. To me it looks like we are not defining a lattice here. 😄

File: core/src/main/scala/lattice/updater.scala [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+  implicit def latticeToUpdater[T](implicit lattice: Lattice[T]): Updater[T] =
+    new AggregationUpdater[T] {
+      override def join(v1: T, v2: T): T = lattice.join(v1, v2)
+
+      override def bottom: T = lattice.bottom
+    }
+
+  def pair[T](implicit lattice: Lattice[T]): Updater[(T, T)] = latticeToUpdater(Lattice.pair(lattice))
+
+  def pair[T](implicit updater: Updater[T]): Updater[(T, T)] = new Updater[(T, T)] with PartialOrderingWithBottom[(T, T)] {
```
Comment (phaller) [inline core/src/main/scala/lattice/updater.scala]: Why are there two `pair` `def`s?

File: core/src/main/scala/lattice/updater.scala [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+    override def bottom: (T, T) =
+      (updater.asInstanceOf[PartialOrderingWithBottom[T]].bottom, updater.asInstanceOf[PartialOrderingWithBottom[T]].bottom)
+
+    override def lteq(x: (T, T), y: (T, T)): Boolean =
+      updater.asInstanceOf[PartialOrderingWithBottom[T]].lteq(x._1, y._1) && updater.asInstanceOf[PartialOrderingWithBottom[T]].lteq(x._2, y._2)
+  }
+}
+
+final case class LatticeViolationException[D](current: D, next: D) extends IllegalStateException(
```
Comment (phaller) [inline core/src/main/scala/lattice/updater.scala]: We should probably rename this to `NotMonotonicException` or something similar, because its main use is within subclasses of `MonotonicUpdater`.

File: core/src/test/scala/cell/LazySuite.scala [hunk 16]
```diff
@@ -2,15 +2,15 @@ package cell
...
+import lattice.{ DefaultKey, Lattice, StringIntKey, StringIntUpdater, Updater }
 import org.scalatest.FunSuite
 
 import scala.concurrent.Await
 import scala.concurrent.duration._
 
 class LazySuite extends FunSuite {
 
-  implicit val stringIntLattice: Lattice[Int] = new StringIntLattice
+  implicit val stringIntUpdatere: Updater[Int] = new StringIntUpdater
```
Comment (phaller) [inline core/src/test/scala/cell/LazySuite.scala]: `stringIntUpdatere` -> `stringIntUpdater`

File: core/src/test/scala/cell/base.scala [hunk 17]
```diff
@@ -1,20 +1,20 @@
...
 
 import org.scalatest.FunSuite
-import java.util.concurrent.{ ConcurrentHashMap, CountDownLatch }
+import java.util.concurrent.CountDownLatch
 
 import scala.util.{ Failure, Success }
 import scala.concurrent.{ Await, Promise }
 import scala.concurrent.duration._
-import lattice.{ DefaultKey, Key, Lattice, LatticeViolationException, StringIntKey, StringIntLattice }
+import lattice._
```
Comment (phaller) [inline core/src/test/scala/cell/base.scala]: Not sure. 😄

File: core/src/test/scala/cell/base.scala [hunk 18]
```diff
@@ -1163,9 +1163,9 @@ class BaseSuite extends FunSuite {
   test("putFinal: result passed to callbacks") {
     val latch = new CountDownLatch(1)
 
-    implicit val setLattice = new Lattice[Set[Int]] {
-      def join(curr: Set[Int], next: Set[Int]) = curr ++ next
-      def empty = Set.empty[Int]
+    implicit val setUpdater = new AggregationUpdater[Set[Int]] {
+      def join(curr: Set[Int], v2: Set[Int]): Set[Int] = curr ++ v2
+      def bottom = Set.empty[Int]
```
Comment (phaller) [inline core/src/test/scala/cell/base.scala]: This should definitely remain a lattice (because it is!), and instead the implicit conversion should convert the lattice to the required updater.

File: core/src/test/scala/cell/base.scala [hunk 19]
```diff
@@ -1190,9 +1190,9 @@ class BaseSuite extends FunSuite {
   test("putNext: result passed to callbacks") {
     val latch = new CountDownLatch(1)
 
-    implicit val setLattice = new Lattice[Set[Int]] {
-      def join(curr: Set[Int], next: Set[Int]) = curr ++ next
-      def empty = Set.empty[Int]
+    implicit val setUpdater = new AggregationUpdater[Set[Int]] {
+      def join(curr: Set[Int], v2: Set[Int]): Set[Int] = curr ++ v2
+      def bottom = Set.empty[Int]
```
Comment (phaller) [inline core/src/test/scala/cell/base.scala]: Ditto.

File: core/src/test/scala/cell/base.scala [hunk 20]
```diff
@@ -1553,9 +1553,9 @@ class BaseSuite extends FunSuite {
   }
 
   test("if exception-throwing tasks should still run quiescent handlers") {
-    val intMaxLattice = new Lattice[Int] {
-      def join(current: Int, next: Int) = math.max(current, next)
-      def empty = 0
+    val intMaxLattice = new MonotonicUpdater[Int] {
+      override def lteq(x: Int, y: Int): Boolean = x <= y
+      def bottom = 0
```
Comment (phaller) [inline core/src/test/scala/cell/base.scala]: As before, this should remain a lattice, and the implicit conversion should turn it into an updater.

File: core/src/test/scala/cell/base.scala [hunk 21]
```diff
@@ -2300,18 +2286,18 @@ class BaseSuite extends FunSuite {
   }
 
   test("whenComplete: called at most once") {
-    val intMaxLattice = new Lattice[Int] {
-      def join(current: Int, next: Int): Int = math.max(current, next)
-      def empty = 0
+    val intMaxUpdater = new MonotonicUpdater[Int] {
+      def lteq(current: Int, next: Int): Boolean = current <= next
+      def bottom = 0
```
Comment (phaller) [inline core/src/test/scala/cell/base.scala]: Should remain a lattice.

File: core/src/test/scala/cell/base.scala [hunk 22]
```diff
@@ -2300,18 +2286,18 @@ class BaseSuite extends FunSuite {
...
     val key = new lattice.DefaultKey[Int]
 
     val pool = new HandlerPool
 
     for (_ <- 1 to 100) {
       val latch = new CountDownLatch(1)
-      val completer1 = CellCompleter[key.type, Int](key)(intMaxLattice, pool)
-      val completer2 = CellCompleter[key.type, Int](key)(intMaxLattice, pool)
+      val completer1 = CellCompleter[key.type, Int](key)(intMaxUpdater, pool)
+      val completer2 = CellCompleter[key.type, Int](key)(intMaxUpdater, pool)
```
Comment (phaller) [inline core/src/test/scala/cell/base.scala]: Should remain lattices.

File: core/src/test/scala/opal/ImmutabilityLattice.scala [hunk 23]
```diff
@@ -31,17 +31,31 @@ case object Immutable extends Immutability
 object Immutability {
 
   implicit object ImmutabilityLattice extends Lattice[Immutability] {
-    override def join(current: Immutability, next: Immutability): Immutability = {
-      if (<=(next, current)) current
-      else next
+    override def join(v1: Immutability, v2: Immutability): Immutability = {
+      if (<=(v2, v1)) v1
```
Comment (phaller) [inline core/src/test/scala/opal/ImmutabilityLattice.scala]: Should use infix syntax.

File: core/src/main/scala/lattice/Lattice.scala (hunk 13 above)
Comment (JanKoelzer) [reply core/src/main/scala/lattice/Lattice.scala]: You are right, join() should not throw this exception. But then, we need to remove `trivial()` as we cannot trivially create a lattice from T <: Null, there is a top value missing. Are you OK with this?

File: core/src/test/scala/opal/ImmutabilityLattice.scala (hunk 23 above)
Comment (JanKoelzer) [reply core/src/test/scala/opal/ImmutabilityLattice.scala]: I have renamed this to lteq to match the naming of PartialOrderingWithBottom.

**Comment by JanKoelzer** [comment]
I have addresses all requested changes, except for now I have not renamed updater.scala to Updater.scala, as this seems to violate http://docs.scala-lang.org/style/files.html#multi-unit-files Are you sure about this?

Secondly, I have removed Lattice.trivial() and instead added PartialOrderingWithBottom.trivial() as this is the most we can get, if there is no "trivial" top element.

**Comment by phaller** [comment]
OK, let's keep `updater.scala` then.
